pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -320,84 +354,93 @@ func getTaskRunName(taskRunsStatus map[string]*v1alpha1.PipelineRunTaskRunStatus
-	if !startTime.IsZero() && pipelineTimeout != nil {
-		timeout := pipelineTimeout.Duration
-		runtime := time.Since(startTime.Time)
-		if runtime > timeout {
-			logger.Infof("PipelineRun %q has timed out(runtime %s over %s)", prName, runtime, timeout)
-
-			timeoutMsg := fmt.Sprintf("PipelineRun %q failed to finish within %q", prName, timeout.String())
-			return &apis.Condition{
-				Type:    apis.ConditionSucceeded,
-				Status:  corev1.ConditionFalse,
-				Reason:  ReasonTimedOut,
-				Message: timeoutMsg,
-			}
+func GetPipelineConditionStatus(pr *v1alpha1.PipelineRun, state PipelineRunState, logger *zap.SugaredLogger, dag *v1alpha1.DAG) *apis.Condition {
+	// We have 4 different states here:
+	// 1. Timed out -> Failed
+	// 2. Any one TaskRun has failed - >Failed. This should change with #1020 and #1023
+	// 3. All tasks are done or are skipped (i.e. condition check failed).-> Success
+	// 4. A Task or Condition is running right now  or there are things left to run -> Running
+
+	if pr.IsTimedOut() {
+		return &apis.Condition{
+			Type:    apis.ConditionSucceeded,
+			Status:  corev1.ConditionFalse,
+			Reason:  ReasonTimedOut,
+			Message: fmt.Sprintf("PipelineRun %q failed to finish within %q", pr.Name, pr.Spec.Timeout.String()),
 		}
 	}
-	allFinished := true
-	for _, rprt := range state {
-		if rprt.TaskRun == nil {
-
-			if rprt.ResolvedConditionChecks == nil {
-				logger.Infof("TaskRun %s doesn't have a Status, so PipelineRun %s isn't finished", rprt.TaskRunName, prName)
-				allFinished = false
-				continue
-			}
-			if !rprt.ResolvedConditionChecks.IsDone() {
-				logger.Infof("ConditionChecks for TaskRun %s in progress, so PipelineRun %s isn't finished", rprt.TaskRunName, prName)
-				allFinished = false
-				continue
-			}
-			if rprt.ResolvedConditionChecks.IsSuccess() {
-				logger.Infof("ConditionChecks for TaskRun %s successful but TaskRun doesn't have a Status, so PipelineRun %s isn't finished", rprt.TaskRunName, prName)
-				allFinished = false
-				continue
-			}
 
-			logger.Info("ConditionChecks for TaskRun %s failed, so PipelineRun %s might be finished unless other TaskRuns are still running", rprt.TaskRunName, prName)
-			continue
-		}
-		c := rprt.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-		if c == nil {
-			logger.Infof("TaskRun %s doesn't have a condition, so PipelineRun %s isn't finished", rprt.TaskRunName, prName)
-			allFinished = false
-			continue
-		}
-		logger.Infof("TaskRun %s status : %v", rprt.TaskRunName, c.Status)
-		// If any TaskRuns have failed, we should halt execution and consider the run failed
-		if c.Status == corev1.ConditionFalse && rprt.IsDone() {
-			logger.Infof("TaskRun %s has failed, so PipelineRun %s has failed, retries done: %b", rprt.TaskRunName, prName, len(rprt.TaskRun.Status.RetriesStatus))
+	// A single failed task mean we fail the pipeline
+	for _, rprt := range state {
+		if rprt.IsFailure() { //IsDone ensures we have crossed the retry limit
```

Comment:
I'm a bit unclear what the `IsDone` comment is referring to here. Did you mean `IsFailure ensures we have crossed...`?

Reply:
Yeah old comment. Will fix